src/sage/schemes/hyperelliptic_curves_smooth_model/hyperelliptic_constructor.py

```diff
@@ -0,0 +1,279 @@
+        sage: HK = H.change_ring(K)
+        sage: HK.points_at_infinity()
+        [(1 : alpha : 0), (1 : -alpha - 1 : 0)]
+        sage: HK.is_split()
+        True
+
+    The construction of hyperelliptic curves is supported over different fields. The correct class is chosen automatically::
+
+        sage: F = FiniteField(13)
+        sage: S.<x> = PolynomialRing(F)
+        sage: HF = HyperellipticCurveSmoothModel(x^5 + x^4 + x^3 + x^2 + x + 1, x^3 + x); HF
+        Hyperelliptic Curve over Finite Field of size 13 defined by y^2 + (x^3 + x)*y = x^5 + x^4 + x^3 + x^2 + x + 1
+        sage: type(HF)
+        <class 'sage.schemes.hyperelliptic_curves_smooth_model.hyperelliptic_g2.HyperellipticCurveSmoothModel_g2_finite_field_with_category'>
+        sage: Q5 = Qp(5,10)
+        sage: T.<x> = Q5[]
+        sage: H5 = HyperellipticCurveSmoothModel(x^7 + 1); H5
+        Hyperelliptic Curve over 5-adic Field with capped relative precision 10 defined by y^2 = x^7 + 1 + O(5^10)
+        sage: type(H5)
+        <class 'sage.schemes.hyperelliptic_curves_smooth_model.hyperelliptic_padic_field.HyperellipticCurveSmoothModel_padic_field_with_category'>
+
+    The input polynomials need not be monic::
+
+        sage: R.<x> = QQ[]
+        sage: HyperellipticCurveSmoothModel(3*x^5+1)
+        Hyperelliptic Curve over Rational Field defined by y^2 = 3*x^5 + 1
+
+    The polynomials `f` and `h` need to define a smooth curve of genus at
+    least one. In particular polynomials defining elliptic curves are
+    allowed as input::
+
+        sage: E = HyperellipticCurveSmoothModel(x^3+1)
+        sage: E.genus()
+        1
+        sage: HyperellipticCurveSmoothModel(x)
+        Traceback (most recent call last):
+        ...
+        ValueError: arguments f = x and h = 0 must define a curve of genus at least one.
+
+    The following polynomials define a singular curve and are
+    not allowed as input::
+
+        sage: C = HyperellipticCurveSmoothModel(x^6 + 2*x - 1, 2*x - 2)
+        Traceback (most recent call last):
+        ...
+        ValueError: singularity in the provided affine patch
+
+    The constructor accepts a distinguished point as an argument::
+
+        sage: C = HyperellipticCurveSmoothModel(x^5 + x + 2, distinguished_point=(1, -2))
+        sage: C.distinguished_point()
+        (1 : -2 : 1)
+
+    """
+
+    # ---------------------------
+    # Internal Helper functions
+    # ---------------------------
+
+    def __genus(f, h):
+        """
+        Helper function to compute the genus of a hyperelliptic curve
+        defined by `y^2 + h(x)y = f(x)`.
+        """
+        # Some classes still have issues with degrees returning `int`
+        # rather than Sage Integer types
+        df = Integer(f.degree())
+        dh_2 = 2 * Integer(h.degree())
+        if dh_2 < df:
+            return (df - 1) // 2
+        return (dh_2 - 1) // 2
+
+    def __check_no_affine_singularities(f, h):
+        """
```

Comment:
```suggestion
        r"""
```

It technically doesn't matter here since these functions don't show up in the docs, but using `r"""` instead of `"""` reduces the amount of escape sequences needed for some LaTeX expressions, and so we prefer it.